### PR TITLE
Add natives to check for NULL_VECTOR and NULL_STRING

### PR DESCRIPTION
**Pull request by peace-maker**
Until now the only use for the `NULL_VECTOR` and `NULL_STRING` pubvars was to pass them to C++ natives for them to be able to skip some work if the callee isn't interested in the results. Plugins themselves couldn't use them for that purpose or anything else at all.
These patches add support for plugins to use them in the same manner as an extension could.

`IsNullVector` and `IsNullString` can be used to check if the `NULL_VECTOR` or `NULL_STRING` was passed as a function/forward argument.

`IsNativeParamNullVector` and `IsNativeParamNullString` can be used to see if someone passed the NULL pubvars when calling your native.

`Call_PushNullVector` and `Call_PushNullString` allow you to call any function or forward and pass the NULL pubvars, which can be checked with the above natives.
Because of the need for the local address instead of the physical one we can't use `IPluginContext::GetNullRef` for these two natives, so the pubvars had to be looked up manually again without caching. If this is a concern we could expose the cached local address too in SourcePawn, but I tried not to change the VM here.

### CODE REVIEW

File: core/logic/ForwardSys.cpp [hunk 1]
```diff
@@ -273,15 +273,53 @@ int CForward::Execute(cell_t *result, IForwardFilter *filter)
...
+					if (param->isnull && type == Param_Array)
+					{
+						IPluginRuntime *runtime = func->GetParentRuntime();
+						uint32_t null_vector_idx;
+						err = runtime->FindPubvarByName("NULL_VECTOR", &null_vector_idx);
+
+						if (!err)
+						{
+							cell_t null_vector;
+							err = runtime->GetPubvarAddrs(null_vector_idx, &null_vector, nullptr);
```
Comment (KyleSanderson) [inline core/logic/ForwardSys.cpp]: Do these need to be different? Can't we have a global null here?

File: core/logic/ForwardSys.cpp [hunk 2]
```diff
@@ -273,15 +273,53 @@ int CForward::Execute(cell_t *result, IForwardFilter *filter)
 				 */
 				if (type == Param_String)
 				{
-					err = func->PushStringEx((char *)param->byref.orig_addr, param->byref.cells, param->byref.sz_flags, param->byref.flags);
+					// If NULL_STRING was pushed, push the reference to the pubvar of the callee instead.
+					if (param->isnull)
```
Comment (KyleSanderson) [inline core/logic/ForwardSys.cpp]: Can we eval the pointer to the array and see if it's null or not instead of maintaining/adding this extra layer?

File: core/logic/smn_fakenatives.cpp [hunk 3]
```diff
@@ -439,5 +491,7 @@ REGISTER_NATIVES(nativeNatives)
 	{"SetNativeArray",			SetNativeArray},
 	{"SetNativeCellRef",		SetNativeCellRef},
 	{"SetNativeString",			SetNativeString},
+	{"IsNativeParamNullVector",	IsNativeParamNullVector},
```
Comment (KyleSanderson) [inline core/logic/smn_fakenatives.cpp]: I'm not certain we should add these, operators are much more natural here in other languages; can we get anywhere close here? Do these need to be different if we maintain a single null?

File: core/logic/ForwardSys.cpp (hunk 2 above)
Comment (peace-maker) [inline core/logic/ForwardSys.cpp]: That's probably a better idea. Would you leave the `CForward::PushNullVector` methods or just allow passing null to PushArray(Ex) and PushString(Ex) (while ignoring the flags)? PushString already allows for NULL, but I don't think that's sane.

File: core/logic/smn_fakenatives.cpp (hunk 3 above)
Comment (peace-maker) [inline core/logic/smn_fakenatives.cpp]: What do you mean by operators?
We could add a single native for this one, since it only checks the parameter number, but the other natives have to be different to be able to take a `const char[]` and a `const float vec[3]` for the pubvars. And for consistency having two seperate natives for `NULL_STRING` and `NULL_VECTOR` respectively like the `IsNullVector` natives seems more intuitive for the user?

File: core/logic/ForwardSys.cpp (hunk 1 above)
Comment (peace-maker) [inline core/logic/ForwardSys.cpp]: They have to be different because of the IsNull[String|Vector] signature. The arguments have be different to check for the pubvar, which could be passed around in local calls too.

**Comment by InstantMuffin** [comment]
Finally this is happening

File: core/logic/ForwardSys.cpp (hunk 2 above)
Comment (dvander) [inline core/logic/ForwardSys.cpp]: This seems fine to me as-is, though the nested |ifs| all over this function are gross, you could factor out a helper if you wanted.

**Comment by dvander** [comment]
Looks good, if you want to do that minor clean-up let me know, otherwise I will merge as-is.

**Comment by peace-maker** [comment]
Cleanup done 🐤

**Comment by InstantMuffin** [comment]
https://img.devrant.io/devrant/rant/r_535939_oThZB.gif please

File: core/logic/ForwardSys.cpp [hunk 4]
```diff
@@ -400,6 +440,7 @@ int CForward::PushCell(cell_t cell)
 		m_params[m_curparam].pushedas = Param_Cell;
 	}
 
+	m_params[m_curparam].isnull = false;
 	m_params[m_curparam++].val = cell;
```
Comment (asherkin) [inline core/logic/ForwardSys.cpp]: Drive-by comment, I find this pattern absolutely terrifying.

Comment (peace-maker) [inline core/logic/ForwardSys.cpp]: You could clean it up in another PR ♿️
